drivers: kscan: Add driver for XPT2046

SeppoTakalo opened the pull request:
Add driver for Xptek XPT2046 resistive touch controller on SPI.
Only interrupt driven mode supported, does not do polling.

Datasheet: https://www.waveshare.com/w/upload/b/b0/XPT2046.pdf
Screen that I'm using to test this: http://www.lcdwiki.com/3.5inch_SPI_Module_ILI9488_SKU:MSP3520

thedjnK commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,276 @@
+	struct xpt2046_data *data = dev->data;
+
+	if (!callback) {
+		LOG_ERR("Callback is null");
+		return -EINVAL;
```
What if a user wants to remove a registered callback?

SeppoTakalo replied:
`kscan_disable_callback(const struct device *dev)`

This is equivalent of what other drivers seem to do.

thedjnK commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,276 @@
+	}
+
+	gpio_init_callback(&data->int_gpio_cb, xpt2046_isr_handler, BIT(config->int_gpio.pin));
+
```
`reads` is not validated either here or DTS - what happens if someone sets it to 0?

SeppoTakalo replied:
I think this is a valid point.
I added some `BUILD_ASSERTS()` at the bottom of the file, so that values have at least some sense.

thedjnK reviewed (commented):
Looks much better than the previous submitted one! Haven't tried it but will do sometime

SeppoTakalo commented on drivers/kscan/Kconfig.xpt2046:
```diff
@@ -0,0 +1,11 @@
+	select SPI
+	help
+	  Enable driver for Xptek XPT2046 resistive touch panel.
+	  This driver is very similar to ADS7843.
```
There seem to be difference in channel numbering and XPT2046 seem to be able to report the touch pressure.
Otherwise, this same driver could be used, but I have no devices to try this, so I did not try to add support for both.

SeppoTakalo commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,276 @@
+
+struct xpt2046_config {
+	/** I2C bus. */
+	const struct spi_dt_spec bus;
```
Thanks. Fixed.

SeppoTakalo commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,276 @@
+	const struct spi_buf_set rx_bufs = {.buffers = &rxb, .count = 1};
+
+	/* Measure 3 times and calculate average */
+	int_fast16_t rounds = config->reads;
```
I changed this to `int` so it does not cause confusion. There was no real reason, other that it is probably so small number that 16 bits would be enough, but let compiler choose the fastest type...

SeppoTakalo commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,276 @@
+	const struct spi_buf_set tx_bufs = {.buffers = &txb, .count = 1};
+	const struct spi_buf_set rx_bufs = {.buffers = &rxb, .count = 1};
+
+	/* Measure 3 times and calculate average */
```
Thanks. Updated.

SeppoTakalo commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,276 @@
+
+	bool pressed = meas.z > config->threshold;
+
+	/* Don't send any other than "pressed" events.
+	 * releasing seem to cause just random noise
```
I refactored the driver a bit so that as long as the PENIRQ is down, assume that user is still pressing. And keep checking this on a worker-queue event. Once the pressure is released, send the "released" event.

This seem to work fine for LVGL.. with one exception: Dragging does not work, because I do not re-read values from SPI. I cannot decide, whether to spend enormous amount of SPI traffic to support dragging, or just be stable with button presses.

bbilas commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,291 @@
+	const struct xpt2046_config *config = dev->config;
+	struct xpt2046_data *data = dev->data;
+
+	if (!device_is_ready(config->bus.bus)) {
```
```suggestion
	if (!spi_is_ready(&config->bus)) {
```

SeppoTakalo replied:
fixed.

bbilas commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,291 @@
+	k_work_init(&data->work, xpt2046_work_handler);
+	k_work_init_delayable(&data->dwork, xpt2046_release_handler);
+
+	if (!device_is_ready(config->int_gpio.port)) {
```
```suggestion
	if (!gpio_is_ready_dt(&config->int_gpio)) {
```

SeppoTakalo replied:
fixed.

SeppoTakalo committed:
drivers: kscan: Add driver for XPT2046

Add driver for Xptek XPT2046 resistive touch controller on SPI.
Only interrupt driven mode supported, does not do polling.

Signed-off-by: Seppo Takalo <seppo.takalo@iki.fi>

bbilas commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,291 @@
+ * Each follow-up command interleaves with previous conversion.
+ * So first command starts at byte 0. Second command starts at byte 2.
+ */
+static uint8_t tbuf[9] = {
```
Shouldn't it be const?

teburd commented:
@bbilas @thedjnK @albertofloyd please take a look

bbilas commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,291 @@
+	.disable_callback = xpt2046_disable_callback,
+};
+
+#define DT_DRV_COMPAT xptek_xpt2046
```
Please move it to the top of the file (as it's usually done in the rest of drivers).

SeppoTakalo replied:
Fixed.

bbilas commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,291 @@
+{
+	int ret = spi_transceive_dt(bus, tx, rx);
+
+	if (ret) {
```
You are mixing the `if (ret)` with `if (ret <0)`. Please unify it (personally I prefer `if (ret < 0)` over `if (ret)` but as I know there is no guideline for that in Zephyr).

SeppoTakalo replied:
Fixed.

bbilas commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,291 @@
+	meas->z += CONVERT_U16(buf, 1) + 4096 - CONVERT_U16(buf, 3);
+	meas->x += CONVERT_U16(buf, 5);
+	meas->y += CONVERT_U16(buf, 7);
+	return 0;
```
Please add nl before return.

SeppoTakalo replied:
Fixed.

bbilas commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,291 @@
+	struct xpt2046_data *data = CONTAINER_OF(dw, struct xpt2046_data, dwork);
+	struct xpt2046_config *config = (struct xpt2046_config *)data->dev->config;
+
+	if (data->pressed && data->enabled) {
```
I prefer to avoid nested conditions. You can use something like below but no strong objections from my side :)

<pre>
if (!data->pressed && !data->enabled) {
    return;
}

/* Check if touch is still pressed */
if (gpio_pin_get_dt(&config->int_gpio) == 0) {
	data->pressed = false;
	data->callback(data->dev, data->last_y, data->last_y, false);
} else {
	/* Re-check later */
	k_work_reschedule(&data->dwork, K_MSEC(10));
}
</pre>

SeppoTakalo replied:
Thanks for the suggestion.

(Actually, the inverted boolean operation convert && to ||, so it is `!data->pressed || !data->enabled`)

bbilas replied:
Right, my bad.

bbilas commented on drivers/kscan/kscan_xpt2046.c:
```diff
@@ -0,0 +1,294 @@
+	LOG_DBG("%s: enable cb", dev->name);
+	data->enabled = true;
+	gpio_add_callback(config->int_gpio.port, &data->int_gpio_cb);
+	return 0;
```
Unfortunately, I've not noticed that one before. Would you mind fixing that by adding an `nl` to all places where it is missing?  Otherwise LGTM!

SeppoTakalo replied:
Sure.
Thanks for being pedantic.
